src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,668 @@
+    pmdaAddLabels(lp, "{\"indom_name\":\"per gpu\"}");
+    return 1;
+  default:
+    break;
+  }
+  return 0;
+}
+
+static int amdgpu_label(int ident, int type, pmLabelSet **lpp, pmdaExt *pmda) {
+  switch (type) {
+  case PM_LABEL_INDOM:
+    amdgpu_labelInDom((pmInDom)ident, lpp);
+    break;
+  default:
+    break;
+  }
+  return pmdaLabel(ident, type, lpp, pmda);
+}
+
+/**
+ * Initializes the path to the help file for this PMDA.
+ */
+static void initializeHelpPath() {
+  int sep = pmPathSeparator();
+  pmsprintf(mypath, sizeof(mypath),
+            "%s%c"
+            "amdgpu"
+            "%c"
+            "help",
+            pmGetConfig("PCP_PMDAS_DIR"), sep, sep);
+}
+
+void __PMDA_INIT_CALL amdgpu_init(pmdaInterface *dp) {
+  if (isDSO) {
+    initializeHelpPath();
+    pmdaDSO(dp, PMDA_INTERFACE_7, "amdgpu DSO", mypath);
+  }
+
+  if (dp->status != 0)
+    return;
+
+  if (!drm_initialized) {
+    setup_gcard_indom();
+    drm_initialized = 1;
+  }
+
+  dp->version.seven.instance = amdgpu_instance;
+  dp->version.seven.fetch = amdgpu_fetch;
+  dp->version.seven.label = amdgpu_label;
+  pmdaSetFetchCallBack(dp, amdgpu_fetchCallBack);
+  pmdaSetLabelCallBack(dp, amdgpu_labelCallBack);
+
+  pmdaInit(dp, indomtab, sizeof(indomtab) / sizeof(indomtab[0]), metrictab,
+           sizeof(metrictab) / sizeof(metrictab[0]));
+}
+
+static void amdgpu_timer(int sig, void *ptr) {
+  (void)sig;
+  (void)ptr;
+  autorefresh = 1;
+}
+
+static void amdgpu_main(pmdaInterface *dispatch) {
+  fd_set readyfds, fds;
+  int pmcdfd, maxfd = 0;
+
+  if ((pmcdfd = __pmdaInFd(dispatch)) < 0)
+    exit(1);
+  if (pmcdfd > maxfd)
+    maxfd = pmcdfd;
+
+  FD_ZERO(&fds);
+  FD_SET(pmcdfd, &fds);
+
+  /* arm interval timer */
+  if (autorefresh == 1 && __pmAFregister(&interval, NULL, amdgpu_timer) < 0) {
+    pmNotifyErr(LOG_ERR, "registering event interval handler");
+    exit(1);
+  }
+
+  for (;;) {
+    memcpy(&readyfds, &fds, sizeof(readyfds));
+    int nready = select(maxfd + 1, &readyfds, NULL, NULL, NULL);
+    if (pmDebugOptions.appl2)
+      pmNotifyErr(LOG_DEBUG, "select: nready=%d autorefresh=%d", nready,
+                  autorefresh);
+    if (nready < 0) {
+      if (neterror() != EINTR) {
+        pmNotifyErr(LOG_ERR, "select failure: %s", netstrerror());
+        exit(1);
+      } else if (autorefresh == 0) {
+        continue;
+      }
+    }
+
+    __pmAFblock();
+    if (nready > 0 && FD_ISSET(pmcdfd, &readyfds)) {
+      if (pmDebugOptions.appl0)
+        pmNotifyErr(LOG_DEBUG, "processing pmcd PDU [fd=%d]", pmcdfd);
+      if (__pmdaMainPDU(dispatch) < 0) {
+        __pmAFunblock();
+        exit(1); /* fatal if we lose pmcd */
+      }
+      if (pmDebugOptions.appl0)
+        pmNotifyErr(LOG_DEBUG, "completed pmcd PDU [fd=%d]", pmcdfd);
+    }
+    if (autorefresh > 0) {
+      autorefresh = 0;
+      for (int i = 0;i < AMDGPU_REFRESHER_COUNT;i++) {
+	  pmNotifyErr(LOG_ERR, "Refreshing %d", i);
```

Comment:
Too verbose by default here, I think this is going to end up in a log file once every second?

Reply:
I need to cleanup, at the end of the day, you were faster than me to review these changes :)